src/hasSubObject.ts

```diff
@@ -1,7 +1,10 @@
-import type { Simplify } from "type-fest";
+import type { Merge } from "type-fest";
+import type { Brand } from "./internal/types";
 import { isDeepEqual } from "./isDeepEqual";
 import { purry } from "./purry";
 
+export type HasSubObjectBrand = Brand<"hasSubObject">;
```

Comment:
We made a big effort to ensure that our types treat symbol props properly, but the basic types built into TypeScript could still create some confusing edge cases.

https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAQgJQIIDkAiMC8MDKBPAWwCMQAbACgHIiAnAQzABNKBKAbgChRJY4tdCJCtVacu4aDAAa-AN4wA2olRoAugC54AGhgALAKalSITQEYATAGYYAXxh0IMbtE7PYATTmLl6DTABERP46BkYmMBbWdg5OElBiHFB4AA760qb8SakgAGbSCgDW+ni5MFn6pVKqHAD0NTANAHoA-ByJKWnuGdjlpe6FxaW9ee6qnHUNMC1AA

Using a symbol key **and value** here makes sense. Symbols are great when you need to differentiate between intents when passing objects around; they make it harder to override or accidentally use the wrong values.

Reply:
@eranhirsch Makes sense, thanks for the example. I will update it to use symbol for the value. Something like this:

```ts
declare const hasSubObjectBrandId: unique symbol;

export type HasSubObjectBrand = Brand<hasSubObjectBrandId>
```

But before I do this, I have another question. I realized that using that single brand type won't handle similar case in nested conditions. For example:

```ts
const obj = { a: 5, b: 5 };

if (hasSubObject(obj, { a: 5 })) {
  obj; // { a: number; b: number; } & HasSubObjectBrand

  if (hasSubObject(obj, { b: 5 })) {
    obj; // { a: number; b: number; } & HasSubObjectBrand
  } else {
    obj; // never <--
  }
} else {
  obj; // { a: number; b: number; }
}
```

It can be solved by having several unique brand types for `hasSubObject` and by recursively checking what brand type `Merge<T, S>` extends:

```ts
declare const hasSubObjectBrandId: unique symbol;
declare const hasSubObjectBrandId2: unique symbol;

export type HasSubObjectBrand = {
  readonly [hasSubObjectBrandId]: typeof hasSubObjectBrandId;
};
// Note: naming isn't final
export type HasSubObjectBrand2 = {
  readonly [hasSubObjectBrandId2]: typeof hasSubObjectBrandId2;
};

export function hasSubObject<T, S extends Partial<T>>(
  data: Merge<T, S> | T,
  subObject: S,
): data is Merge<T, S> extends HasSubObjectBrand
  ? Merge<T, S> & HasSubObjectBrand2
  : Merge<T, S> & HasSubObjectBrand;
  
const obj = { a: 5, b: 5 };
const subObj = { b: 5 };

if (hasSubObject(obj, subObj)) {
  obj; // { a: number; b: number; } & HasSubObjectBrand

  if (hasSubObject(obj, subObj)) {
    obj; // { a: number; b: number; } & HasSubObjectBrand & HasSubObjectBrand2 
  } else {
    obj; // { a: number; b: number; } & HasSubObjectBrand
  }
} else {
  obj; // { a: number; b: number; }
}
```

I'm not sure if we need to handle that case. What do you think? Should we handle it? If so, how deep we should handle this?

Reply:
i think it's fine if we don't handle this case; the workaround is

```ts
if (hasSubObject(obj, { ...subObj1, ...subObj2 }) {
  // ...
} else if (hasSubObject(obj, subObj1)) {
  // ...
} else if (hasSubObject(obj, subObj2)) {
  // ...
} else {
  // ...
}
```

right? if there's a complaint in the future, like if someone wants to nest more than two hasSubObjects, then we can look into this again, but for now i think it's fine?

Reply:
> if there's a complaint in the future, like if someone wants to nest more than two hasSubObjects, then we can look into this again, but for now i think it's fine?

@cjquines Agree. There are different workarounds and the case probably isn't popular so it shouldn't be a blocker.

Reply:
@eranhirsch @cjquines Updated PR to use symbol for the value of the brand type.